src/main/java/org/jenkinsci/plugins/workflow/libs/LibraryCachingConfiguration.java

```diff
@@ -83,29 +83,43 @@ public static FilePath getGlobalLibrariesCacheDir() {
     }
 
     @Extension public static class DescriptorImpl extends Descriptor<LibraryCachingConfiguration> {
-        public FormValidation doClearCache(@QueryParameter String name) throws InterruptedException {
+        public FormValidation doClearCache(@QueryParameter String name, @QueryParameter String cachedLibraryRef) throws InterruptedException {
             Jenkins.get().checkPermission(Jenkins.ADMINISTER);
-
+            String cacheDirName = null;
             try {
                 if (LibraryCachingConfiguration.getGlobalLibrariesCacheDir().exists()) {
-                    for (FilePath libraryNamePath : LibraryCachingConfiguration.getGlobalLibrariesCacheDir().list("*-name.txt")) {
-                        // Libraries configured in distinct locations may have the same name. Since only admins are allowed here, this is not a huge issue, but it is probably unexpected.
-                        String cacheName;
-                        try (InputStream stream = libraryNamePath.read()) {
-                            cacheName = IOUtils.toString(stream, StandardCharsets.UTF_8);
-                        }
-                        if (libraryNamePath.readToString().equals(name)) {
-                            FilePath libraryCachePath = LibraryCachingConfiguration.getGlobalLibrariesCacheDir()
-                                    .child(libraryNamePath.getName().replace("-name.txt", ""));
-                            libraryCachePath.deleteRecursive();
-                            libraryNamePath.delete();
+                    outer: for (FilePath libraryCache : LibraryCachingConfiguration.getGlobalLibrariesCacheDir().listDirectories()) {
+                        for (FilePath libraryNamePath : libraryCache.list("*-name.txt")) {
+                            if (libraryNamePath.readToString().startsWith(name + "@")) {
+                                FilePath libraryCachePath = libraryNamePath.getParent();
+                                if (libraryCachePath != null) {
+                                    FilePath versionCachePath = new FilePath(libraryCachePath, libraryNamePath.getName().replace("-name.txt", ""));
+                                    if (cachedLibraryRef != null && !cachedLibraryRef.equals("")) {
+                                        if (libraryNamePath.readToString().equals(name + "@" + cachedLibraryRef)) {
+                                            cacheDirName = name + "@" + cachedLibraryRef;
+                                            libraryNamePath.delete();
+                                            versionCachePath.deleteRecursive();
+                                            break outer;
+                                        }
+                                    } else {
+                                        cacheDirName = name;
+                                        libraryCachePath.deleteRecursive();
```

Comment:
When #3 is merged you will not be able to delete like this. You must delete each version separately to ensure it is not just getting used by any other operation (e.g. cache was expired and is getting refreshed, a build is just copying it).

Reply:
Sure. I will integrate your changes depending on which gets merged first.